frontend/src/metabase/lib/api.js

```diff
@@ -376,6 +376,79 @@ export class Api extends EventEmitter {
         }
       });
   }
+
+  /**
+   /**
+   * @typedef {{
+   *    method: "GET" | "POST";
+   *    url: string;
+   *    options: {
+   *      headers?: Record<string, string>;
+   *      hasBody: boolean;
+   *    } & Record<string, unknown>;
+   * }} OnBeforeRequestHandlerData
+   *
+   * @typedef {(data: OnBeforeRequestHandlerData) => (Promise<void | OnBeforeRequestHandlerData>)} OnBeforeRequestHandler
+   *
+   * @typedef {{
+   *   key: string;
+   *   handler: OnBeforeRequestHandler
+   * }} OnBeforeRequestHandlerDescriptor
+   *
+   * @param data {OnBeforeRequestHandlerData}
+   * @return data {Promise<OnBeforeRequestHandlerData>}
+   */
+  async apiRequestManipulationMiddleware(data) {
+    let { method, url, options } = data;
+
+    /**
+     * Handlers order is important.
+     * Handlers are executed in order and each handler uses the data returned by a previous handler.
+     * @type {OnBeforeRequestHandler[]}
+     */
+    const handlers = [];
+
+    if (isEmbeddingSdk()) {
+      if (
+        PLUGIN_EMBEDDING_SDK.onBeforeRequestHandlers.getOrRefreshSessionHandler
+      ) {
+        // A handler that set's `getOrRefreshSession` logic for SDK
+        handlers.push(
+          PLUGIN_EMBEDDING_SDK.onBeforeRequestHandlers
+            .getOrRefreshSessionHandler,
+        );
+      }
+    }
```

Comment:
This seems better, and I assume future handlers will also be pushed in this same function, so we can just read here what all the possible manipulations are, and under what conditions they occur 👍